File: vtk-rs/libvtkrs/src/vtk_parametric_klein.cpp
#include <cstring>
#include <sstream>

#include "cxx.h"
#include "vtk_parametric_klein.h"
#include "vtk_parametric_klein.rs.h"

#include <vtkParametricKlein.h>
#include <vtkParametricFunction.h>

vtkParametricKlein* vtk_parametric_klein_new() {
    vtkParametricKlein* obj = vtkParametricKlein::New();
    if (!obj) {
        throw std::runtime_error("Failed to create vtkParametricKlein");
    }
    return obj;
}

void vtk_parametric_klein_delete(vtkParametricKlein& klein) {
    klein.Delete();
}

vtkParametricFunction* parametric_klein_as_parametric_function(vtkParametricKlein& klein) {
    return static_cast<vtkParametricFunction*>(&klein);
}

File: vtk-rs/src/vtk_parametric_klein.rs
#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("vtk_parametric_klein.h");
        include!("vtk_parametric_function.h");

        type vtkParametricKlein;
        type vtkParametricFunction;

        fn vtk_parametric_klein_new() -> *mut vtkParametricKlein;
        fn vtk_parametric_klein_delete(ptr: Pin<&mut vtkParametricKlein>);
        unsafe fn parametric_klein_as_parametric_function(
            klein: Pin<&mut vtkParametricKlein>
        ) -> *mut vtkParametricFunction;
    }
}

crate::define_object!(
    "https://vtk.org/doc/nightly/html/classvtkParametricKlein.html",
    @name ParametricKlein, ffi::vtkParametricKlein,
    @new ffi::vtk_parametric_klein_new,
    @delete ffi::vtk_parametric_klein_delete,
    @inherit vtkObjectBase
);

impl ParametricKlein {
    /// Get a pointer to the base vtkParametricFunction for use with ParametricFunctionSource
    pub fn as_parametric_function(&mut self) -> *mut std::ffi::c_void {
        unsafe {
            ffi::parametric_klein_as_parametric_function(self.ptr.as_mut()) as *mut std::ffi::c_void
        }
    }
}
